subxt/src/backend/mod.rs

```diff
@@ -325,9 +325,395 @@ pub enum TransactionStatus<Hash> {
 pub struct StorageResponse {
     /// The key.
     pub key: Vec<u8>,
     /// The associated value.
     pub value: Vec<u8>,
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    mod legacy {
+        use std::{
+            collections::{HashMap, VecDeque},
+            sync::Arc,
+        };
+        use tokio::sync::{mpsc, Mutex};
+
+        use super::rpc::{RpcClient, RpcClientT};
+        use crate::backend::rpc::RawRpcSubscription;
+        use crate::backend::BackendExt;
+        use crate::{
+            backend::{
+                legacy::rpc_methods::Bytes, legacy::rpc_methods::RuntimeVersion,
+                legacy::LegacyBackend, StorageResponse,
+            },
+            error::RpcError,
+        };
+        use futures::StreamExt;
+        use serde::Serialize;
+        use serde_json::value::RawValue;
+        use subxt_core::{config::DefaultExtrinsicParams, Config};
+        type RpcResult<T> = Result<T, RpcError>;
+        type Item = RpcResult<String>;
+
+        struct StorageEntry {
+            items: HashMap<Vec<u8>, VecDeque<Item>>,
+        }
+
+        trait Storage {
+            fn new() -> Self;
+            fn push<I: Serialize>(&mut self, key: Vec<u8>, item: RpcResult<I>);
+
+            fn pop(&mut self, key: Vec<u8>) -> Item;
+        }
+
+        impl Storage for StorageEntry {
+            fn new() -> Self {
+                StorageEntry {
+                    items: HashMap::new(),
+                }
+            }
+            fn push<I: Serialize>(&mut self, key: Vec<u8>, item: RpcResult<I>) {
+                let item = item.map(|x| serde_json::to_string(&x).unwrap());
+                match self.items.entry(key) {
+                    std::collections::hash_map::Entry::Occupied(v) => v.into_mut().push_back(item),
+                    std::collections::hash_map::Entry::Vacant(e) => {
+                        e.insert(VecDeque::from([item]));
+                    }
+                }
+            }
+
+            fn pop(&mut self, key: Vec<u8>) -> Item {
+                self.items.get_mut(&key).unwrap().pop_front().unwrap()
+            }
+        }
+
+        struct Data<R: Storage> {
+            request: R,
+            subscription: mpsc::Receiver<RpcResult<Vec<Item>>>,
+        }
+
+        struct MockRpcClientStorage {
+            data: Arc<Mutex<Data<StorageEntry>>>,
+        }
+
+        impl RpcClientT for MockRpcClientStorage {
+            fn request_raw<'a>(
+                &'a self,
+                method: &'a str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> super::rpc::RawRpcFuture<'a, Box<serde_json::value::RawValue>> {
+                Box::pin(async move {
+                    match method {
+                        "state_getStorage" => {
+                            let mut data = self.data.lock().await;
+                            let key = match serde_json::to_value(params.unwrap()).unwrap() {
```

Comment:
I think jsonpsee is already in the dependency tree and you could simplify this code to use.

```rust
   let rpc_params = jsonrpsee::types::Params::new(params.map(|p| p.get());
   let key: sp_core::Bytes = rpc_params.one().unwrap();
```

Perhaps move to helper function to make it easier to read and follow

Reply:
done as suggested